stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/repository/domain/message/attachment/UploadAttachmentsWorker.kt

```diff
@@ -0,0 +1,192 @@
+package io.getstream.chat.android.offline.repository.domain.message.attachment
+
+import android.content.Context
+import android.webkit.MimeTypeMap
+import androidx.work.CoroutineWorker
+import androidx.work.OneTimeWorkRequestBuilder
+import androidx.work.WorkManager
+import androidx.work.WorkerParameters
+import androidx.work.workDataOf
+import io.getstream.chat.android.client.ChatClient
+import io.getstream.chat.android.client.call.await
+import io.getstream.chat.android.client.models.Attachment
+import io.getstream.chat.android.client.uploader.ProgressTracker
+import io.getstream.chat.android.client.uploader.ProgressTrackerFactory
+import io.getstream.chat.android.client.uploader.StreamCdnImageMimeTypes
+import io.getstream.chat.android.client.uploader.toProgressCallback
+import io.getstream.chat.android.client.utils.SyncStatus
+import io.getstream.chat.android.offline.ChatDomain
+import io.getstream.chat.android.offline.ChatDomainImpl
+import io.getstream.chat.android.offline.extensions.isPermanent
+import io.getstream.chat.android.offline.repository.RepositoryFacade
+import java.io.File
+import io.getstream.chat.android.client.utils.Result as CallResult
+
+internal class UploadAttachmentsWorker(
+    appContext: Context,
+    workerParams: WorkerParameters,
+) : CoroutineWorker(appContext, workerParams) {
+
+    val repository: RepositoryFacade
+        get() {
+            return (ChatDomain.instance() as ChatDomainImpl).repos
+        }
+
+    override suspend fun doWork(): Result {
+        val channelType: String = inputData.getString(DATA_CHANNEL_TYPE)!!
+        val channelId: String = inputData.getString(DATA_CHANNEL_ID)!!
+        val messageId = inputData.getString(DATA_MESSAGE_ID)!!
+
+        val attachmentEntities: List<AttachmentToUploadEntity> = repository
+            .selectAttachmentsToUploadForMessageId(messageId)
+            .filter { it.syncStatus == SyncStatus.SYNC_NEEDED }
+
+        if (runAttemptCount > RETRY_COUNT) {
+            cleanup(attachmentEntities)
+            return Result.failure()
+        }
+
+        return try {
+            var batchUploaded = true
+
+            attachmentEntities.forEach { entity ->
+                repository.updateAttachmentsToUpload(entity.copy(syncStatus = SyncStatus.IN_PROGRESS))
+
+                val result = uploadAttachment(channelType, channelId, entity)
+                if (result.isSuccess) {
+                    val augmentedEntity = result.data()
+                    repository.updateAttachmentsToUpload(augmentedEntity.copy(syncStatus = SyncStatus.COMPLETED))
```

Comment:
Should we remove old entries from the AttachmentsToUpload table somewhere? 🤔

Reply:
Yes, probably. But how do we know that some upload has failed.

Reply:
I will remove it in my part of the code `sendMessage`